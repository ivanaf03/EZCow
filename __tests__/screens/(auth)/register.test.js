import React from "react";

import {render, fireEvent, waitFor} from "@testing-library/react-native";
import { router } from 'expo-router';
import { Alert } from "react-native";

import Register from "../../../app/(auth)/register";

jest.mock("expo-router", () => ({
    router: {
        push: jest.fn(),
    },
}));

jest.mock("../../../app/model/users", () => ({
    insertUser: jest.fn(),
}));

jest.spyOn(Alert, "alert");

jest.mock("@fortawesome/react-native-fontawesome", () => ({
    FontAwesomeIcon: () => null
}));

describe("Register", () => {

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it("should render correctly", () => {
        const tree = render(<Register />);
        expect(tree).toMatchSnapshot();
    });

    it("sign up link should navigate to login screen", () => {
        const {getByTestId} = render(<Register />);
        const signInLink = getByTestId("sign-in-link");
        fireEvent.press(signInLink);
        expect(router.push).toHaveBeenCalledWith("login");
    });

    it("shows an alert when email is not valid", () => {
        const {getByPlaceholderText, getAllByPlaceholderText, getByTestId} = render(<Register />);
        const nameInput = getByPlaceholderText("Nombre");
        fireEvent.changeText(nameInput, "A");
        const emailInput = getByPlaceholderText("Email");
        fireEvent.changeText(emailInput, "A");
        const passwordInput = getAllByPlaceholderText("Contraseña")[0];
        fireEvent.changeText(passwordInput, "12345678");
        const confirmPasswordInput = getAllByPlaceholderText("Contraseña")[1];
        fireEvent.changeText(confirmPasswordInput, "12345678");
        const signUpButton = getByTestId("sign-up-button");
        fireEvent.press(signUpButton);
        expect(Alert.alert).toHaveBeenCalledWith('Error', 'El email no es válido.');
    });

    it("shows an alert when password is too short", () => {
        const {getByPlaceholderText, getAllByPlaceholderText, getByTestId} = render(<Register />);
        const nameInput = getByPlaceholderText("Nombre");
        fireEvent.changeText(nameInput, "A");
        const emailInput = getByPlaceholderText("Email");
        fireEvent.changeText(emailInput, "test@example.com");
        const passwordInput = getAllByPlaceholderText("Contraseña")[0];
        fireEvent.changeText(passwordInput, "123456");
        const confirmPasswordInput = getAllByPlaceholderText("Contraseña")[1];
        fireEvent.changeText(confirmPasswordInput, "123456");
        const signUpButton = getByTestId("sign-up-button");
        fireEvent.press(signUpButton);
        expect(Alert.alert).toHaveBeenCalledWith('Error', 'La contraseña debe tener al menos 8 caracteres.');
    });

    it("shows an alert when passwords are not the same", () => {
        const {getByPlaceholderText, getAllByPlaceholderText, getByTestId} = render(<Register />);
        const nameInput = getByPlaceholderText("Nombre");
        fireEvent.changeText(nameInput, "A");
        const emailInput = getByPlaceholderText("Email");
        fireEvent.changeText(emailInput, "test@example.com");
        const passwordInput = getAllByPlaceholderText("Contraseña")[0];
        fireEvent.changeText(passwordInput, "12345678");
        const confirmPasswordInput = getAllByPlaceholderText("Contraseña")[1];
        fireEvent.changeText(confirmPasswordInput, "123456789");
        const signUpButton = getByTestId("sign-up-button");
        fireEvent.press(signUpButton);
        expect(Alert.alert).toHaveBeenCalledWith('Error', 'Las contraseñas no coinciden.');
    });

    it("shows an alert when email is already registered", () => {    
        const {getByPlaceholderText, getAllByPlaceholderText, getByTestId} = render(<Register />);
        const nameInput = getByPlaceholderText("Nombre");
        fireEvent.changeText(nameInput, "A");
        const emailInput = getByPlaceholderText("Email");
        fireEvent.changeText(emailInput, "test@example.com");
        const passwordInput = getAllByPlaceholderText("Contraseña")[0];
        fireEvent.changeText(passwordInput, "12345678");
        const confirmPasswordInput = getAllByPlaceholderText("Contraseña")[1];
        fireEvent.changeText(confirmPasswordInput, "12345678");
        const signUpButton = getByTestId("sign-up-button");
        fireEvent.press(signUpButton);
        waitFor(() => expect(Alert.alert).toHaveBeenCalledWith('Error', 'Error al registrar. Inténtalo de nuevo. Asegúrate de que no tengas ya una cuenta.'));
    });

    it("works when email, password and confirm password are correct", async() => {
        const {getByPlaceholderText, getAllByPlaceholderText, getByTestId} = render(<Register />);
        const nameInput = getByPlaceholderText("Nombre");
        fireEvent.changeText(nameInput, "A");
        const emailInput = getByPlaceholderText("Email");
        fireEvent.changeText(emailInput, "test@example.com");
        const passwordInput = getAllByPlaceholderText("Contraseña")[0];
        fireEvent.changeText(passwordInput, "12345678");
        const confirmPasswordInput = getAllByPlaceholderText("Contraseña")[1];
        fireEvent.changeText(confirmPasswordInput, "12345678");
        const signUpButton = getByTestId("sign-up-button");
        fireEvent.press(signUpButton);
        await waitFor(() => expect(router.push).toHaveBeenCalledWith('home'));
    });
});