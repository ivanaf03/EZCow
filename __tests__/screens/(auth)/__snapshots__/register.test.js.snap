// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register should render correctly 1`] = `
<RCTScrollView
  style={
    [
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      {
        "backgroundColor": "#1c1c1c",
      },
    ]
  }
>
  <View>
    <Text
      style={
        [
          {
            "marginTop": 16,
          },
          {
            "marginLeft": 16,
            "marginRight": 16,
          },
          {
            "color": "#ffffff",
          },
          {
            "fontSize": 36,
            "lineHeight": 40,
          },
          {
            "fontFamily": "Nunito-Bold",
          },
        ]
      }
    >
      Registrarse
    </Text>
    <View
      style={
        [
          {
            "marginLeft": 16,
            "marginRight": 16,
          },
          {
            "borderLeftWidth": 2,
          },
          {
            "borderRightWidth": 2,
          },
          {
            "borderBottomColor": "#6bc0eb",
            "borderLeftColor": "#6bc0eb",
            "borderRightColor": "#6bc0eb",
            "borderTopColor": "#6bc0eb",
          },
          {
            "borderBottomLeftRadius": 16,
            "borderBottomRightRadius": 16,
            "borderTopLeftRadius": 16,
            "borderTopRightRadius": 16,
          },
          {
            "paddingBottom": 16,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 16,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "marginLeft": 8,
            },
          ]
        }
      >
        <View
          style={
            [
              {
                "width": "100%",
              },
            ]
          }
        >
          <View
            style={
              [
                {
                  "marginTop": -4,
                },
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#ffffff",
                  },
                  {
                    "fontSize": 14,
                    "lineHeight": 20,
                  },
                  {
                    "fontFamily": "Nunito-Medium",
                  },
                  [
                    {
                      "marginTop": 4,
                    },
                  ],
                ]
              }
            >
              Nombre de usuario
              :
            </Text>
            <TextInput
              onChangeText={[Function]}
              placeholder="Nombre"
              style={
                [
                  {
                    "paddingBottom": 8,
                    "paddingLeft": 8,
                    "paddingRight": 8,
                    "paddingTop": 8,
                  },
                  {
                    "backgroundColor": "#ffffff",
                  },
                  {
                    "borderBottomWidth": 2,
                    "borderLeftWidth": 2,
                    "borderRightWidth": 2,
                    "borderTopWidth": 2,
                  },
                  {
                    "borderBottomColor": "#2254ba",
                    "borderLeftColor": "#2254ba",
                    "borderRightColor": "#2254ba",
                    "borderTopColor": "#2254ba",
                  },
                  {
                    "borderBottomLeftRadius": 16,
                    "borderBottomRightRadius": 16,
                    "borderTopLeftRadius": 16,
                    "borderTopRightRadius": 16,
                  },
                  {
                    "fontFamily": "Nunito-Medium",
                  },
                  [
                    {
                      "marginTop": 4,
                    },
                  ],
                ]
              }
            />
          </View>
        </View>
        <View
          style={
            [
              {
                "marginTop": 40,
              },
              {
                "width": "100%",
              },
            ]
          }
        >
          <View
            style={
              [
                {
                  "marginTop": -4,
                },
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#ffffff",
                  },
                  {
                    "fontSize": 14,
                    "lineHeight": 20,
                  },
                  {
                    "fontFamily": "Nunito-Medium",
                  },
                  [
                    {
                      "marginTop": 4,
                    },
                  ],
                ]
              }
            >
              Email
              :
            </Text>
            <TextInput
              onChangeText={[Function]}
              placeholder="Email"
              style={
                [
                  {
                    "paddingBottom": 8,
                    "paddingLeft": 8,
                    "paddingRight": 8,
                    "paddingTop": 8,
                  },
                  {
                    "backgroundColor": "#ffffff",
                  },
                  {
                    "borderBottomWidth": 2,
                    "borderLeftWidth": 2,
                    "borderRightWidth": 2,
                    "borderTopWidth": 2,
                  },
                  {
                    "borderBottomColor": "#2254ba",
                    "borderLeftColor": "#2254ba",
                    "borderRightColor": "#2254ba",
                    "borderTopColor": "#2254ba",
                  },
                  {
                    "borderBottomLeftRadius": 16,
                    "borderBottomRightRadius": 16,
                    "borderTopLeftRadius": 16,
                    "borderTopRightRadius": 16,
                  },
                  {
                    "fontFamily": "Nunito-Medium",
                  },
                  [
                    {
                      "marginTop": 4,
                    },
                  ],
                ]
              }
            />
          </View>
        </View>
        <View
          style={
            [
              {
                "marginTop": 40,
              },
              {
                "width": "100%",
              },
            ]
          }
        >
          <View
            style={
              [
                {
                  "marginTop": -4,
                },
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#ffffff",
                  },
                  {
                    "fontSize": 14,
                    "lineHeight": 20,
                  },
                  {
                    "fontFamily": "Nunito-Light",
                  },
                  [
                    {
                      "marginTop": 4,
                    },
                  ],
                ]
              }
            >
              Contraseña
              :
            </Text>
            <View
              style={
                [
                  {
                    "flexDirection": "row",
                  },
                  {
                    "alignItems": "center",
                  },
                  [
                    {
                      "marginTop": 4,
                    },
                  ],
                ]
              }
            >
              <TextInput
                onChangeText={[Function]}
                placeholder="Contraseña"
                secureTextEntry={true}
                style={
                  [
                    {
                      "paddingBottom": 8,
                      "paddingLeft": 8,
                      "paddingRight": 8,
                      "paddingTop": 8,
                    },
                    {
                      "marginRight": 8,
                    },
                    {
                      "backgroundColor": "#ffffff",
                    },
                    {
                      "borderBottomWidth": 2,
                      "borderLeftWidth": 2,
                      "borderRightWidth": 2,
                      "borderTopWidth": 2,
                    },
                    {
                      "borderBottomColor": "#2254ba",
                      "borderLeftColor": "#2254ba",
                      "borderRightColor": "#2254ba",
                      "borderTopColor": "#2254ba",
                    },
                    {
                      "borderBottomLeftRadius": 16,
                      "borderBottomRightRadius": 16,
                      "borderTopLeftRadius": 16,
                      "borderTopRightRadius": 16,
                    },
                    {
                      "flexBasis": "0%",
                      "flexGrow": 1,
                      "flexShrink": 1,
                    },
                    {
                      "fontFamily": "Nunito-Medium",
                    },
                  ]
                }
              />
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                testID="eye-icon"
              />
            </View>
          </View>
        </View>
        <View
          style={
            [
              {
                "marginTop": 40,
              },
              {
                "width": "100%",
              },
            ]
          }
        >
          <View
            style={
              [
                {
                  "marginTop": -4,
                },
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#ffffff",
                  },
                  {
                    "fontSize": 14,
                    "lineHeight": 20,
                  },
                  {
                    "fontFamily": "Nunito-Light",
                  },
                  [
                    {
                      "marginTop": 4,
                    },
                  ],
                ]
              }
            >
              Confirmar contraseña
              :
            </Text>
            <View
              style={
                [
                  {
                    "flexDirection": "row",
                  },
                  {
                    "alignItems": "center",
                  },
                  [
                    {
                      "marginTop": 4,
                    },
                  ],
                ]
              }
            >
              <TextInput
                onChangeText={[Function]}
                placeholder="Contraseña"
                secureTextEntry={true}
                style={
                  [
                    {
                      "paddingBottom": 8,
                      "paddingLeft": 8,
                      "paddingRight": 8,
                      "paddingTop": 8,
                    },
                    {
                      "marginRight": 8,
                    },
                    {
                      "backgroundColor": "#ffffff",
                    },
                    {
                      "borderBottomWidth": 2,
                      "borderLeftWidth": 2,
                      "borderRightWidth": 2,
                      "borderTopWidth": 2,
                    },
                    {
                      "borderBottomColor": "#2254ba",
                      "borderLeftColor": "#2254ba",
                      "borderRightColor": "#2254ba",
                      "borderTopColor": "#2254ba",
                    },
                    {
                      "borderBottomLeftRadius": 16,
                      "borderBottomRightRadius": 16,
                      "borderTopLeftRadius": 16,
                      "borderTopRightRadius": 16,
                    },
                    {
                      "flexBasis": "0%",
                      "flexGrow": 1,
                      "flexShrink": 1,
                    },
                    {
                      "fontFamily": "Nunito-Medium",
                    },
                  ]
                }
              />
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                testID="eye-icon"
              />
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "justifyContent": "center",
            },
            {
              "marginTop": 48,
            },
          ]
        }
      >
        <View
          style={
            [
              {
                "width": "90%",
              },
            ]
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "paddingBottom": 16,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "paddingTop": 16,
                },
                {
                  "backgroundColor": "#272727",
                },
                {
                  "borderLeftWidth": 8,
                  "borderRightWidth": 8,
                },
                {
                  "borderBottomWidth": 2,
                  "borderTopWidth": 2,
                },
                {
                  "borderBottomColor": "#2254ba",
                  "borderLeftColor": "#2254ba",
                  "borderRightColor": "#2254ba",
                  "borderTopColor": "#2254ba",
                },
                {
                  "borderBottomLeftRadius": 24,
                  "borderBottomRightRadius": 24,
                  "borderTopLeftRadius": 24,
                  "borderTopRightRadius": 24,
                },
                {
                  "alignItems": "center",
                },
                {
                  "justifyContent": "center",
                },
              ]
            }
            testID="sign-up-button"
          >
            <Text
              style={
                [
                  {
                    "fontFamily": "Nunito-SemiBold",
                  },
                  {
                    "color": "#ffffff",
                  },
                  {
                    "fontSize": 20,
                    "lineHeight": 28,
                  },
                  {
                    "textAlign": "center",
                  },
                ]
              }
            >
              Registrarse
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        [
          {
            "marginTop": 32,
          },
          {
            "width": "100%",
          },
          {
            "marginBottom": 56,
          },
          {
            "flexDirection": "row",
          },
          {
            "justifyContent": "center",
          },
          {
            "alignItems": "center",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#ffffff",
            },
            {
              "fontSize": 20,
              "lineHeight": 28,
            },
            {
              "fontFamily": "Nunito-Medium",
            },
            {
              "textAlign": "center",
            },
            {
              "marginRight": 8,
            },
          ]
        }
      >
        Ya tienes una cuenta?
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="sign-in-link"
      >
        <Text
          style={
            [
              {
                "color": "#6bc0eb",
              },
              {
                "fontSize": 20,
                "lineHeight": 28,
              },
              {
                "textAlign": "center",
              },
              {
                "fontFamily": "Nunito-Italic",
              },
              {
                "textDecorationLine": "underline",
              },
            ]
          }
        >
          Iniciar sesión
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
